src/main/java/io/netty/incubator/codec/http3/Http3ControlStreamInboundHandler.java

```diff
@@ -176,4 +204,105 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+
+        AbstractQPackStreamInitializer(int streamType) {
+            this.streamType = streamType;
+        }
+
+        @Override
+        public final void channelActive(ChannelHandlerContext ctx) {
+            // We need to write the streamType into the stream before doing anything else.
+            // See https://tools.ietf.org/html/draft-ietf-quic-http-32#section-6.2.1
+            // Just allocate 8 bytes which would be the max needed.
+            ByteBuf buffer = ctx.alloc().buffer(8);
+            Http3CodecUtils.writeVariableLengthInteger(buffer, streamType);
+            ctx.writeAndFlush(buffer);
+            streamAvailable(ctx);
+            ctx.fireChannelActive();
+        }
+
+        @Override
+        public final void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+            if (evt instanceof ChannelInputShutdownEvent) {
+                // See https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#section-4.2
+                Http3CodecUtils.criticalStreamClosed(ctx);
+            }
+            ctx.fireUserEventTriggered(evt);
+        }
+
+        @Override
+        public void channelInactive(ChannelHandlerContext ctx) {
+            // See https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#section-4.2
+            Http3CodecUtils.criticalStreamClosed(ctx);
+            ctx.fireChannelInactive();
+        }
+
+        protected abstract void streamAvailable(ChannelHandlerContext ctx);
+    }
+
+    private static final class QPackEncoderStreamInitializer extends AbstractQPackStreamInitializer {
+        private static final ClosedChannelException ENCODER_STREAM_INACTIVE =
+                unknownStackTrace(new ClosedChannelException(), ClosedChannelException.class, "channelInactive()");
+        private final QpackEncoder encoder;
+        private final long maxTableCapacity;
+        private final long blockedStreams;
+
+        QPackEncoderStreamInitializer(QpackEncoder encoder, long maxTableCapacity, long blockedStreams) {
+            super(Http3CodecUtils.HTTP3_QPACK_ENCODER_STREAM_TYPE);
+            this.encoder = encoder;
+            this.maxTableCapacity = maxTableCapacity;
+            this.blockedStreams = blockedStreams;
+        }
+
+        @Override
+        protected void streamAvailable(ChannelHandlerContext ctx) {
+            final QuicStreamChannel stream = (QuicStreamChannel) ctx.channel();
+            try {
+                encoder.configureDynamicTable(stream, maxTableCapacity, blockedStreams);
+            } catch (QpackException e) {
+                connectionError(ctx, new Http3Exception(QPACK_ENCODER_STREAM_ERROR,
+                        "Dynamic table configuration failed.", e), true);
+                return;
+            }
+            final Channel parent = ctx.channel().parent();
+            ((Promise<QuicStreamChannel>) Http3.getQPackEncoderStreamAvailable(parent)).trySuccess(stream);
+            Http3.setQPackEncoderStream(parent, stream);
```

Comment:
should we set this first before we notify as a listener may try to use what we set here ?